ran-core/src/main/java/io/ran/ClazzMethod.java

```diff
@@ -83,8 +90,67 @@ public boolean isSynchronized() {
 		return Modifier.isSynchronized(method.getModifiers());
 	}
 
+	private List<Type> getAllGenericTypes() {
+		ArrayList<Type> types = new ArrayList<>();
+		types.add(method.getGenericReturnType());
+		types.addAll(Arrays.asList(method.getGenericParameterTypes()));
+		types.addAll(Arrays.asList(method.getGenericExceptionTypes()));
+		return types.stream().flatMap(type -> {
+			if(type instanceof ParameterizedType) {
+				return Stream.of(((ParameterizedType)type).getActualTypeArguments());
+			} else {
+				return Stream.of(type);
+			}
+		}).collect(Collectors.toList());
+	}
+
+	/**
+	 * @return true when the method either returns or takes generic parameters defined at class-level
+	 */
+	public boolean hasGenericFromClass() {
+		List<Type> types = getAllGenericTypes();
+		for(Type type : types) {
+			if(type instanceof TypeVariable<?>) {
+				TypeVariable<?> typed = (TypeVariable<?>) type;
+				if(typed.getGenericDeclaration() == method.getDeclaringClass())
+					return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * @return true if the method definition has generic parameters at method-level.
+	 */
+	public boolean hasGenericFromMethod() {
+		List<Type> types = getAllGenericTypes();
+		for(Type type : types) {
+			if(type instanceof TypeVariable<?>) {
+				TypeVariable<?> typed = (TypeVariable<?>) type;
+				if(typed.getGenericDeclaration() == method)
+					return true;
+			} else if(type instanceof WildcardType) {
+				//WildcardType typed = (WildcardType) type;
+				//if(typed.getTypeName().equals("?"))
+				return true;
+			}
+		}
+		return false;
+	}
+
 	public Clazz<?> getReturnType() {
-		return Clazz.of(method.getReturnType());
+		Class<?> declaringClass = method.getDeclaringClass();
```

Comment:
I don't see any tests that show that getReturnType still works for non generic types, which the implementation of this suggest may not work correctly. Considering it is used a lot of other places it may have been tested implicitly, but an explicit test would be better to have as well